Various fixes to make clang-tidy happier

larseggert opened the pull request:
This is a little all over the place, but consolidates a bunch of nit fixes to `core` and `platform` that make `clang-tidy` happier. There are still some issues it finds, incl. some zero-pointer accesses, etc.

Are you interested in fixes like this?

nibanks commented on src/core/binding.c:
```diff
@@ -318,13 +318,17 @@ QuicBindingRegisterListener(
 
         if (NewFamily > ExistingFamily) {
             break; // End of possible family matches. Done searching.
-        } else if (NewFamily != ExistingFamily) {
```
Does clang actually complain about this?

ThadHouse replied:
I'm definitely not a big fan of replacing all else ifs with else statements, which it looks like this PR does a lot of.

nibanks replied:
I actually don't mind those changes. After looking at the code a while, I think I might like it better.

larseggert replied:
The `else if` -> `if` changes address the [readability-else-after-return](https://clang.llvm.org/extra/clang-tidy/checks/readability-else-after-return.html) and similar warning. I prefer it, but we can also simply disable those checks.

nibanks commented on src/core/binding.c:
```diff
@@ -1618,7 +1625,6 @@ QuicBindingReceive(
         //
         if (!QuicBindingDeliverDatagrams(Binding, SubChain, SubChainLength)) {
             *ReleaseChainTail = SubChain;
-            ReleaseChainTail = SubChainTail;
```
I guess I understand that this is unnecessary, but I fear that if any other code added after this that does manipulation of the chain might result in bugs if we remove this. What do you think?

larseggert replied:
Leave it, but comment it out, explaining why? Or `// NOLINT` the line?

nibanks replied:
I'd say `// NOLINT` the line.

nibanks commented on src/core/configuration.c:
```diff
@@ -358,7 +358,7 @@ MsQuicConfigurationLoadCredential(
 void
 QuicConfigurationTraceRundown(
-    _In_ QUIC_CONFIGURATION* Configuration
+    _In_ QUIC_CONFIGURATION* Configuration // NOLINT
```
What does the `// NOLINT` do here?

larseggert replied:
It stops a warning that `Configuration` is unused, because on Linux `QuicTraceEvent` is a no-op.

nibanks replied:
I'd prefer `UNREFERENCED_PARAMETER` instead, since that's what we do everywhere else.

larseggert replied:
I can make that change (also elsewhere). But don't you think it's confusing, since the parameter is actually referenced when `QuicTraceEvent` isn't a no-op?

nibanks commented on src/core/connection.c:
```diff
@@ -2389,14 +2389,18 @@ QuicConnValidateTransportParameterDraft27CIDs(
-        } else if (
+        }
+
+        if (
             memcmp(
```
```suggestion
        if (memcmp(
```

nibanks commented on src/core/connection.c:
```diff
@@ -4911,7 +4917,6 @@ QuicConnRecvDatagrams(
             Batch,
             Cipher,
             &RecvState);
-        BatchCount = 0;
```
Again, I don't like removing lines like this that could result in bugs in the future.

nibanks replied:
I think I'd rather NOLINT these type of things.

nibanks commented on src/core/connection.c:
```diff
@@ -5416,10 +5421,12 @@ QuicConnParamSet(
             QUIC_ALLOC_NONPAGED(BufferLength, QUIC_POOL_CLOSE_REASON);
 
         if (Connection->CloseReasonPhrase != NULL) {
```
Combine this if and the buffer if statements?

nibanks commented on src/core/crypto.c:
```diff
@@ -201,7 +201,7 @@ QuicCryptoInitialize(
 
     if (QUIC_FAILED(Status)) {
-        for (uint8_t i = 0; i < QUIC_PACKET_KEY_COUNT; ++i) {
+        for (size_t i = 0; i < QUIC_PACKET_KEY_COUNT; ++i) {
```
Why?

ThadHouse replied:
My guess is since its indexing, it prefers the size that can index into an array. The generated assembly is the same, as the variable is stuck in a register anyway, so I think this is ok. Less chance for accidental rollover bugs in the future, even though I know QUIC_PACKET_KEY_COUNT likely will not change.

larseggert replied:
Because `QUIC_PACKET_KEY_COUNT` is an `enum`, i.e., an `int`. Could also use `int`.

nibanks commented on src/core/crypto_tls.c:
```diff
@@ -155,6 +155,9 @@ QuicCryptoTlsReadSniExtension(
     _Inout_ QUIC_NEW_CONNECTION_INFO* Info
     )
 {
+
+    UNREFERENCED_PARAMETER(Connection);
+
```
```suggestion
    UNREFERENCED_PARAMETER(Connection);
```

nibanks commented on src/core/library.c:
```diff
@@ -206,7 +206,6 @@ MsQuicLibraryInitialize(
             LibraryStorageOpenFailed,
             "[ lib] Failed to open global settings, 0x%x",
             Status);
-        Status = QUIC_STATUS_SUCCESS; // Non-fatal, as the process may not have access
```
We should at least keep the comment.

nibanks commented on src/core/range.c:
```diff
@@ -278,7 +280,7 @@ QuicRangeAddRange(
                     QuicRangeCompare(&Key, Sub) == 0) {
                 --i;
             }
-            Sub = QuicRangeGet(Range, i);
+            QuicRangeGet(Range, i);
```
This doesn't look right, does it?

nibanks replied:
Well, I guess it's never being read, so the whole line should be removed?

ThadHouse replied:
Its unconditionally overwritten on line 294, so I understand why its flagged. If that is the case, is this call then needed?

larseggert replied:
I wasn't sure if the call has side effects. I guess not.

ThadHouse commented on src/core/configuration.c:
```diff
@@ -408,34 +408,24 @@ QuicConfigurationParamGet(
-
-    switch (Param) {
-    case QUIC_PARAM_CONFIGURATION_SETTINGS:
+    if (Param == QUIC_PARAM_CONFIGURATION_SETTINGS) {
```
Isn't this a place where we're likely to add more options to the switch statement in the future? 

Also, I know multi returns is fine in C, but its definitely changing from what is mostly our standard. Definitely just want to double check that we're ok with this.

nibanks replied:
I am fine with these changes. If we do need to add more in the future, it's easy enough to change then.

ThadHouse commented on src/core/connection.c:
```diff
@@ -1612,7 +1612,7 @@ QuicConnTryClose(
         }
 
-        if (Flags & QUIC_CLOSE_APPLICATION) {
+        if (Flags & QUIC_CLOSE_APPLICATION) { // NOLINT
```
Why is it flagging so many if statements with // NOLINT? This flags one _maybe_ I can understand, but the one below on line 2036 is odd to me.

larseggert replied:
Because `QUIC_CLOSE_APPLICATION` is a (signed) `int`, and bitops on signed values are undefined.

larseggert replied:
The one on 2036 (and many others) is because both paths of the `if` have the same code - `QuicTraceEvent` is a no-op on Linux.

I can comment out this check instead?

nibanks replied:
The logging logic here is important. What's the correct way to fix this logic? Why is it treating `QUIC_CLOSE_APPLICATION` as a signed int?

larseggert replied:
The issue here is that both paths of the `if` are identical, because `QuicTraceEvent` is a no-op. That is what the `NOLINT` suppresses.

(I was wrong on the bitops thing; must have looked at some other code piece.)

nibanks commented on src/core/stream.c:
```diff
@@ -572,7 +572,6 @@ QuicStreamParamGet(
 
         if (!Stream->Flags.Started ||
             !Stream->Flags.LocalCloseAcked) {
-            Status = QUIC_STATUS_INVALID_STATE;
```
Instead of removing this, add a `break` afterwards.

nibanks commented on src/platform/datapath_linux.c:
```diff
@@ -664,19 +664,21 @@ QuicDataPathPopulateTargetAddress(
-    } else if (AddrInfo->ai_addr->sa_family == AF_INET) {
+        Address->Ipv6 = *SockAddrIn6;
+        return;
+
```
nit: remove extra newline

nibanks commented on src/platform/platform_linux.c:
```diff
@@ -13,16 +13,16 @@
 #include <syslog.h>
-#include <dlfcn.h>
-#include "quic_trace.h"
-#include "quic_platform_dispatch.h"
```
Don't remove this header.

larseggert replied:
Not removed, just reordered

nibanks commented on src/platform/platform_linux.c:
```diff
@@ -146,7 +146,7 @@ QuicPlatformInitialize(
     QUIC_FRE_ASSERT(PlatDispatch != NULL);
 #else
-    RandomFd = open("/dev/urandom", O_RDONLY);
+    RandomFd = open("/dev/urandom", O_RDONLY|O_CLOEXEC);
```
I don't completely understand the intricacies here. Why is `O_CLOEXEC` being added?

Maximus- replied:
Not sure why it was added here by Lars, but it's definitely good hygiene for programs that fork. That being said, I don't know what MsQuics expected behavior is in a process that forks ;P

larseggert replied:
https://clang.llvm.org/extra/clang-tidy/checks/android-cloexec-open.html

Can also just disable this check.

nibanks replied:
I don't have the expertise here so I'll have to you defer to whatever you guys think is best. I believe I originally grabbed this code from stackoverflow.

larseggert replied:
There is IMO no downside to this change. @Maximus- ?

ThadHouse replied:
I worry that this would break macOS when we get to that PR.

larseggert commented on src/platform/datapath_linux.c:
```diff
@@ -13,14 +13,14 @@
 #include <inttypes.h>
 #include <linux/in6.h>
-#include <arpa/inet.h>
-#include "quic_platform_dispatch.h"
```
It's not removed, just reordered

nibanks commented:
Windows build is breaking:
```
D:\a\1\msquic\src\core\sent_packet_metadata.c(66,1): error C2220: the following warning is treated as an error [D:\a\1\msquic\build\windows\x86_schannel\src\core\core.vcxproj]
D:\a\1\msquic\src\core\sent_packet_metadata.c(66,1): warning C4267: 'initializing': conversion from 'size_t' to 'uint16_t', possible loss of data [D:\a\1\msquic\build\windows\x86_schannel\src\core\core.vcxproj]
```

nibanks commented on src/core/connection.c:
```diff
@@ -5811,9 +5812,9 @@ QuicConnParamGet(
-            Stats->Timing.Start = QuicTimeUs64ToPlat(Stats->Timing.Start);
-            Stats->Timing.InitialFlightEnd = QuicTimeUs64ToPlat(Stats->Timing.InitialFlightEnd);
-            Stats->Timing.HandshakeFlightEnd = QuicTimeUs64ToPlat(Stats->Timing.HandshakeFlightEnd);
+            Stats->Timing.Start = QuicTimeUs64ToPlat(Stats->Timing.Start); // cppcheck-suppress selfAssignment
+            Stats->Timing.InitialFlightEnd = QuicTimeUs64ToPlat(Stats->Timing.InitialFlightEnd); // cppcheck-suppress selfAssignment
+            Stats->Timing.HandshakeFlightEnd = QuicTimeUs64ToPlat(Stats->Timing.HandshakeFlightEnd); // cppcheck-suppress selfAssignment
```
What's up with these?

larseggert replied:
`QuicTimeUs64ToPlat` is a no-op, so this is a pure (useless) self-assignment.

nibanks commented on src/core/crypto.c:
```diff
@@ -105,6 +105,7 @@ QuicCryptoInitialize(
             QUIC_MAX_TLS_SERVER_SEND_BUFFER : QUIC_MAX_TLS_CLIENT_SEND_BUFFER;
     uint16_t InitialRecvBufferLength =
         QuicConnIsServer(Connection) ?
+            // cppcheck-suppress duplicateValueTernary
```
Can we just disable this warning/error globally instead?

nibanks commented on src/core/packet.h:
```diff
@@ -307,6 +307,7 @@ QuicPktNumEncode(
     )
 {
     for (uint8_t i = 0; i < PacketNumberLength; i++) {
+        // cppcheck-suppress objectIndex
```
What's up here?

larseggert replied:
The warning is `The address of local variable 'PacketNumber' might be accessed at non-zero index. [objectIndex]`. 
Can also disable this globally?

nibanks replied:
Probably.

nibanks commented on src/core/registration.c:
```diff
@@ -243,37 +243,35 @@ MsQuicRegistrationShutdown(
         QUIC_TRACE_API_REGISTRATION_SHUTDOWN,
         Handle);
 
-    if (Handle && Handle->Type == QUIC_HANDLE_TYPE_REGISTRATION) {
```
Why are you removing this?

larseggert replied:
Because it is asserted to be true in lines 233-234.

nibanks replied:
They are debug asserts. They don't fire in release builds. We need these checks still.

nibanks commented on src/core/send.c:
```diff
@@ -118,7 +118,7 @@ _IRQL_requires_max_(DISPATCH_LEVEL)
 QuicSendQueueFlush(
     _In_ QUIC_SEND* Send,
-    _In_ QUIC_SEND_FLUSH_REASON Reason
+    _In_ QUIC_SEND_FLUSH_REASON Reason // NOLINT
```
I prefer `UNREFERENCED_PARAMETER` instead of `NOLINT` for variables only used in tracing.

larseggert replied:
See above. Am fine with this, but I felt it was confusing, because the variable is referenced when the trace call isn't a no-op.

nibanks replied:
`// NOLINT` isn't going to help cross-platform if logging is disabled, but `UNREFERENCED_PARAMETER` will. BTW, I don't want to worry/care/optimize-for the no-logging scenario since it's not the norm.

nibanks commented on src/core/stream.c:
```diff
@@ -345,7 +345,7 @@ QuicStreamClose(
 void
 QuicStreamTraceRundown(
-    _In_ QUIC_STREAM* Stream
+    _In_ QUIC_STREAM* Stream // NOLINT
```
`UNREFERENCED_PARAMETER`

larseggert commented:
I don't understand what the `cmake` error on Windows is about.

nibanks commented:
> I don't understand what the `cmake` error on Windows is about.

```D:\a\1\msquic\src\core\sent_packet_metadata.c(72,31): warning C4267: 'function': conversion from 'size_t' to 'uint32_t', possible loss of data [D:\a\1\msquic\build\windows\x64_schannel\src\core\core.vcxproj]
```
This error? The problem is that you're passing a `size_t` to a function (`QuicPoolInitialize`) that takes a `uint32_t` without an explicit cast.

larseggert commented:
Ah, so the error is two screens up in the log - cunning! :-)

larseggert commented:
@nibanks I think this is ready. The one failing run seems to be because clog didn't get downloaded for some reason.

nibanks commented on src/core/unittest/FrameTest.cpp:
```diff
@@ -173,7 +173,7 @@ TEST_P(AckFrameTest, DecodeAckFrameFail) {
         Buffer[3] = 0; // ACK range count
         Buffer[4] = 5; // First ACK range
-        for (auto TestValue : {64, 255}) {
+        for (uint8_t TestValue = 64; TestValue <= 255; TestValue++) {
```
Why?

larseggert replied:
Because otherwise it's an int, and then the Windows compiler throws an error about "loss of precision"

ThadHouse replied:
Isn't this going to infinite loop? Since 255 is true, then rollover will happen and the condition will still be true.

nibanks commented:
Looks like you broke some things, because various tests are all failing.

- SpinQuic is hanging
- Perf seems to crash

ThadHouse commented:
The perf crash is the following

```
0000008f`060ff820 00007ff6`c269653d     quicperf!_invoke_watson+0x18
0000008f`060ff850 00007ff6`c2696559     quicperf!_invalid_parameter+0xad
0000008f`060ff890 00007ff6`c2699b63     quicperf!_invalid_parameter_noinfo+0x19
0000008f`060ff8d0 00007ff6`c2699e9e     quicperf!__acrt_stdio_parse_mode<char>+0x2ab
(Inline Function) --------`--------     quicperf!common_openfile+0xc
0000008f`060ff900 00007ff6`c268a880     quicperf!_openfile+0x26
(Inline Function) --------`--------     quicperf!__crt_char_traits<char>::open_file+0xe
0000008f`060ff960 00007ff6`c254b9e1     quicperf!common_fsopen<char>+0x90
0000008f`060ff9a0 00007ff6`c2545d55     quicperf!QuicTlsGenerateSelfSignedCert+0x2d1
(Inline Function) --------`--------     quicperf!QuicPlatGetSelfSignedCert+0xd8
0000008f`060ffa30 00007ff6`c26819dc     quicperf!main+0x475
(Inline Function) --------`--------     quicperf!invoke_main+0x22
0000008f`060ffbd0 00007ff8`e68c4d0e     quicperf!__scrt_common_main_seh+0x10c
0000008f`060ffc10 00007ff8`e7315deb     KERNEL32!BaseThreadInitThunk+0x1e
0000008f`060ffc40 00000000`00000000     ntdll!RtlUserThreadStart+0x2b
```

ThadHouse commented on src/platform/selfsign_openssl.c:
```diff
@@ -216,7 +216,7 @@ QuicTlsGenerateSelfSignedCert(
     Fd = NULL;
 
-    Fd = fopen(CertFileName, "wb");
+    Fd = fopen(CertFileName, "wbe");
```
This is also called from windows, which doesn't accept the e flag. Since the e flag is specific only to gclibc, I would very much prefer this not be in here at all, because otherwise mac will break when we get to it too.

larseggert replied:
Darwin has support for `e`; but I have removed it anyway in the latest push.

larseggert commented:
So I think I fixed the crashes that I introduced. I will note that `spinquic` still crashes after some time, but it also does that when compiled from the current `main` branch. FWIW, here is the stack trace for those crashes:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
#1  0x00007f9bf1de6864 in __GI_abort () at abort.c:79
#2  0x0000565532153fba in quic_bugcheck () at ../src/platform/platform_linux.c:60
#3  0x0000565532152d91 in ServerSpin (Context=<optimized out>) at ../src/tools/spin/spinquic.cpp:625
#4  0x00007f9bf1fb3590 in start_thread (arg=0x7f9beb45e640) at pthread_create.c:463
#5  0x00007f9bf1ed9223 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
#1  0x00007f6efa25d864 in __GI_abort () at abort.c:79
#2  0x0000563c5f37ffba in quic_bugcheck () at ../src/platform/platform_linux.c:60
#3  0x0000563c5f37eeb1 in ServerSpin (Context=<optimized out>) at /usr/include/x86_64-linux-gnu/bits/byteswap.h:37
#4  0x00007f6efa42a590 in start_thread (arg=0x7f6ef38d5640) at pthread_create.c:463
#5  0x00007f6efa350223 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

nibanks commented:
It seems both Windows OpenSSL and Windows miTLS BVTs are failing to run and just timing out (which is incorrectly showing as a pass, btw).

larseggert commented:
Sigh. Is there a way to debug why it hangs based on the CI log?

nibanks commented:
> Sigh. Is there a way to debug why it hangs based on the CI log? I don't have a Windows box or any idea how to replicate this on one.
> 
> (And what's BVT?)

No logs or dumps were uploaded for the tests unfortunately (I think because they're not getting marked as a hard fail for some reason). Also BVT means "Build Verification Test".

larseggert commented:
@nibanks:
```
##[error]There is not enough space on the disk. : 'C:\a\_diag\pages\ab09ace5-9714-4291-b529-adb1a8db6aa3_1.log'
```

larseggert commented:
@nibanks also, is it possible to get more output from the BVT runs? i have really no idea how to debug the hangs.

nibanks commented:
> @nibanks also, is it possible to get more output from the BVT runs? i have really no idea how to debug the hangs.

I'm not sure how to expose build output publicly. There is no reason it needs to be hidden. Is there a way easily to split this PR into smaller changes? Or at least individual commits you can push through the CI one at a time?

larseggert commented:
I can try, but it will be some days until I can

larseggert commented:
Closing this in favor of #1085
